File: src/screen_ocr/src/result_mapper.cpp
#include "screen_ocr/result_mapper.hpp"

#include <cmath>
#include <limits>
#include <regex>
#include <string>

namespace screen_ocr
{
namespace
{

constexpr double kUnknownCovariance = -1.0;

std::optional<double> parse_number(const std::string & text)
{
  if (text.empty() || text == "none") {
    return std::nullopt;
  }

  static const std::regex number_re(R"(([-+]?\d*\.?\d+))");
  std::smatch match;
  if (!std::regex_search(text, match, number_re)) {
    return std::nullopt;
  }

  try {
    return std::stod(match[1].str());
  } catch (const std::exception &) {
    return std::nullopt;
  }
}

geometry_msgs::msg::Vector3 heading_to_magnetic_field(std::optional<double> heading_deg)
{
  geometry_msgs::msg::Vector3 vector;
  if (!heading_deg.has_value()) {
    return vector;
  }

  const double radians = heading_deg.value() * M_PI / 180.0;
  vector.x = std::sin(radians);
  vector.y = std::cos(radians);
  vector.z = 0.0;
  return vector;
}

std::pair<bool, bool> arrow_flags(const std::string & arrow_direction)
{
  std::string direction = arrow_direction;
  for (char & ch : direction) {
    ch = static_cast<char>(std::tolower(static_cast<unsigned char>(ch)));
  }

  const bool left = direction == "left" || direction == "both";
  const bool right = direction == "right" || direction == "both";
  return {left, right};
}

float to_nan_if_missing(const std::optional<double> & value)
{
  if (!value.has_value()) {
    return std::numeric_limits<float>::quiet_NaN();
  }
  return static_cast<float>(value.value());
}

}  // namespace

screen_ocr_msgs::msg::SensorMsg recognition_to_sensor_msg(
  const RecognitionResult & recognition,
  const std_msgs::msg::Header & header)
{
  const auto percent = parse_number(recognition.signal_strength);
  const auto current = parse_number(recognition.pipeline_current);
  const auto depth = parse_number(recognition.burial_depth);

  std::optional<double> heading = recognition.compass_angle_deg;
  if (!heading.has_value()) {
    heading = parse_number(recognition.compass_angle);
  }

  const auto [left_arrow, right_arrow] = arrow_flags(recognition.arrow_direction);

  screen_ocr_msgs::msg::SensorMsg msg;
  msg.header = header;
  msg.magnetic_field = heading_to_magnetic_field(heading);
  msg.magnetic_field_covariance = {
    kUnknownCovariance, 0.0, 0.0,
    0.0, kUnknownCovariance, 0.0,
    0.0, 0.0, kUnknownCovariance,
  };

  msg.signal_strength_percent = to_nan_if_missing(percent);
  msg.signal_strength = percent.has_value()
    ? static_cast<float>(percent.value() / 100.0)
    : std::numeric_limits<float>::quiet_NaN();
  msg.depth_meters = to_nan_if_missing(depth);
  msg.current_milliamps = to_nan_if_missing(current);
  msg.pipeline_heading_degrees = to_nan_if_missing(heading);
  msg.left_arrow = left_arrow;
  msg.right_arrow = right_arrow;
  return msg;
}

}  // namespace screen_ocr

File: src/screen_ocr/screen_ocr/result_mapper.py
"""Map screen OCR recognition output to screen_ocr_msgs/SensorMsg."""

from __future__ import annotations

import math
import re
from typing import Any

from geometry_msgs.msg import Vector3
from screen_ocr_msgs.msg import SensorMsg
from std_msgs.msg import Header


_UNKNOWN_COVARIANCE = -1.0


def _parse_number(text: str) -> float | None:
    if not text or text == "none":
        return None
    match = re.search(r"([-+]?\d*\.?\d+)", text)
    return float(match.group(1)) if match else None


def _heading_to_magnetic_field(heading_deg: float | None) -> Vector3:
    vector = Vector3()
    if heading_deg is None:
        return vector

    radians = math.radians(heading_deg)
    vector.x = math.sin(radians)
    vector.y = math.cos(radians)
    vector.z = 0.0
    return vector


def _default_covariance() -> list[float]:
    return [
        _UNKNOWN_COVARIANCE, 0.0, 0.0,
        0.0, _UNKNOWN_COVARIANCE, 0.0,
        0.0, 0.0, _UNKNOWN_COVARIANCE,
    ]


def _arrow_flags(arrow_direction: str) -> tuple[bool, bool]:
    direction = (arrow_direction or "none").lower()
    left = direction in {"left", "both"}
    right = direction in {"right", "both"}
    return left, right


def recognition_to_sensor_msg(
    recognition: dict[str, Any],
    header: Header,
) -> SensorMsg:
    percent = _parse_number(recognition.get("signal_strength", ""))
    current = _parse_number(recognition.get("pipeline_current", ""))
    depth = _parse_number(recognition.get("burial_depth", ""))
    heading = recognition.get("compass_angle_deg")
    if heading is None:
        heading = _parse_number(recognition.get("compass_angle", ""))

    left_arrow, right_arrow = _arrow_flags(recognition.get("arrow_direction", "none"))

    msg = SensorMsg()
    msg.header = header
    msg.magnetic_field = _heading_to_magnetic_field(heading)
    msg.magnetic_field_covariance = _default_covariance()
    msg.signal_strength_percent = float(percent) if percent is not None else float("nan")
    msg.signal_strength = (
        float(percent) / 100.0 if percent is not None else float("nan")
    )
    msg.depth_meters = float(depth) if depth is not None else float("nan")
    msg.current_milliamps = float(current) if current is not None else float("nan")
    msg.pipeline_heading_degrees = float(heading) if heading is not None else float("nan")
    msg.left_arrow = left_arrow
    msg.right_arrow = right_arrow
    return msg
